utils/s2n_array.c

```diff
@@ -51,26 +51,46 @@ static S2N_RESULT s2n_array_enlarge(struct s2n_array *array, uint32_t capacity)
 
 struct s2n_array *s2n_array_new(uint32_t element_size)
 {
-    struct s2n_blob mem = {0};
+    struct s2n_array *array = s2n_array_new_with_capacity(element_size, S2N_INITIAL_ARRAY_SIZE);
+    PTR_ENSURE_REF(array);
+
+    return array;
+}
+
+struct s2n_array *s2n_array_new_with_capacity(uint32_t element_size, uint32_t capacity)
+{
+    DEFER_CLEANUP(struct s2n_blob mem = {0}, s2n_free);
     PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_array)));
 
     struct s2n_array *array = (void *) mem.data;
 
-    *array = (struct s2n_array) {.mem = {0}, .len = 0, .element_size = element_size};
+    PTR_GUARD_RESULT(s2n_array_init_with_capacity(array, element_size, capacity));
 
-    if (s2n_result_is_error(s2n_array_enlarge(array, S2N_INITIAL_ARRAY_SIZE))) {
-        /* Avoid memory leak if allocation fails */
-        PTR_GUARD_POSIX(s2n_free(&mem));
-        return NULL;
-    }
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
     return array;
 }
 
 S2N_RESULT s2n_array_init(struct s2n_array *array, uint32_t element_size)
 {
     RESULT_ENSURE_REF(array);
 
-    *array = (struct s2n_array){.element_size = element_size};
+    RESULT_GUARD(s2n_array_init_with_capacity(array, element_size, 0));
+
+    RESULT_GUARD(s2n_array_validate(array));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_array_init_with_capacity(struct s2n_array *array, uint32_t element_size, uint32_t capacity)
+{
+    RESULT_ENSURE_REF(array);
+
+    *array = (struct s2n_array){
+        .mem = { 0 },
+        .len = 0,
+        .element_size = element_size,
+    };
+
+    RESULT_GUARD(s2n_array_enlarge(array, capacity));
 
     RESULT_GUARD(s2n_array_validate(array));
```

Comment:
We could argue you don't need to validate again, since s2n_array_enlarge already does it for you. Maybe the rule should be, call validate if you've modified the array directly since the last s2n_array_* call? I don't think we've ever defined it well.

Reply:
Okay yeah, that makes sense!

Reply:
I added more s2n_array_validates in 07091062e2c30dcbb4d6a2f2dc1db3b4aa83899c where this rule wasn't followed.